product_variant_exclusion/models/product.py

```diff
@@ -0,0 +1,174 @@
+                        'attribute_value_ids').ids:
+                    raise UserError(_(
+                        'You cannot delete an attribute line when '
+                        'one of its values is part of an exclusion'))
+        return super(ProductAttributeLine, self).unlink()
+
+
+class ProductVariantExclusion(models.Model):
+
+    _name = 'product.variant.exclusion'
+
+    product_tmpl_id = fields.Many2one('product.template', required=True)
+    attribute_value_ids = fields.Many2many('product.attribute.value',
+                                           required=True)
+
+    def _check_values_attributes_uniqueness(self, value_ids):
+        values = self.env['product.attribute.value'].browse(value_ids)
+        attributes = values.mapped('attribute_id')
+        if len(values) != len(attributes):
+            raise UserError(_('You cannot have multiple values from the '
+                              'same attribute in an exclusion.'))
+
+    @api.model
+    def create(self, vals):
+        value_ids = vals.get('attribute_value_ids')[0][2]
+        if len(value_ids) <= 1:
+            raise UserError(_(
+                'You cannot create an exclusion with only one value !'))
+        template_lines = self.env['product.attribute.line'].search([
+            ('product_tmpl_id', '=', vals.get('product_tmpl_id')),
+            ('value_ids', 'in', value_ids)])
+        for line in template_lines:
+            if len(line.value_ids) <= 1 or set(line.value_ids.ids) <= set(
+                    value_ids):
+                raise UserError(_(
+                    'You cannot create an exclusion for an attribute value '
+                    'which is unique in its attribute line !'))
+        self._check_values_attributes_uniqueness(value_ids)
+        return super(ProductVariantExclusion, self).create(vals)
+
+    @api.multi
+    def write(self, vals):
+        value_ids = vals.get('attribute_value_ids')[0][2]
+        if len(value_ids) <= 1:
+            raise UserError(_(
+                'You cannot modify an exclusion to have only one value !'))
+        self._check_values_attributes_uniqueness(value_ids)
+        return super(ProductVariantExclusion, self).write(vals)
+
+
+class ProductProduct(models.Model):
+
+    _inherit = 'product.product'
+
+    @api.multi
+    def write(self, vals):
+        if self._context.get('create_product_variant'):
+            products_to_write = self.env['product.product']
+            # Case 2: Setting active to True
+            if vals.get('active', False):
+                for product in self:
+                    product_value_ids = product.attribute_value_ids.ids
+                    if product.product_tmpl_id.variant_exclusion_ids:
+                        for excl in product.product_tmpl_id.\
+                                variant_exclusion_ids:
+                            if not set(excl.attribute_value_ids.ids) <= set(
```

Comment:
someone will know better than me, but isn't there a more new API way of doing these recordset comparisons?